File: src/types/sync.ts
/**
 * @file sync.ts
 * @brief 使用了精确类型定义的业务逻辑代码, 为了与数据库和后端代码的命名符合统一使用 snake_case 命名。
 * @author gwj
 * @date 2025年9月30日
 */

import { User } from "@/services/user";

/**
 * 通用服务器响应结构，与 Rust 中的 `ApiResponse<T>` 对应。
 */
export interface ApiResponse<T> {
  success: boolean;
  code: number;
  message: string;
  data?: T | null; // Option<T> 映射为 T | null，并且字段可选
}

/**
 * 网站分组的数据传输对象，与 Rust 中的 `WebsiteGroupDto` 对应。
 */
export interface WebsiteGroupDto {
  uuid: string;
  name: string;
  description: string | null;
  sort_order: number | null;
  is_deleted: number;
  updated_at: string;
}

/**
 * 网站项的数据传输对象，与 Rust 中的 `WebsitesDto` 对应。
 */
export interface WebsitesDto {
  uuid: string;
  group_uuid: string;
  title: string;
  url: string;
  url_lan: string | null;
  default_icon: string | null;
  local_icon_path: string | null;
  background_color: string | null;
  description: string | null;
  sort_order: number | null;
  is_deleted: number;
  updated_at: string;
}

/**
 * 资产分类的数据传输对象，与 Rust 中的 `AssetCategoryDto` 对应。
 */
export interface AssetCategoryDto {
  uuid: string;
  name: string;
  is_default: number;
  is_deleted: number;
  updated_at: string;
}

/**
 * 资产的数据传输对象，与 Rust 中的 `AssetDto` 对应。
 */
export interface AssetDto {
  uuid: string;
  category_uuid: string;
  name: string;
  purchase_date: string;
  price: number;
  expiration_date: string | null;
  description: string | null;
  is_deleted: number;
  updated_at: string;
  brand: string | null;
  model: string | null;
  serial_number: string | null;
  status: string;
  sale_price: number | null;
  sale_date: string | null;
  fees: number | null;
  buyer: string | null;
  notes: string | null;
  realized_profit: number | null;
}

/**
 * 搜索引擎的数据传输对象
 */
export interface SearchEngineDto {
  uuid: string;
  name: string;
  url_template: string;
  default_icon: string | null;
  local_icon_path: string | null;
  is_default: number;
  sort_order: number | null;
  updated_at: string;
}

/**
 * 同步数据包的核心内容，与 Rust 中的 `SyncDataDto` 对应。
 * 注意：Rust 结构体使用了 `rename_all = "camelCase"`，因此属性名使用驼峰式。
 */
export interface SyncDataDto {
  website_groups: WebsiteGroupDto[];
  websites: WebsitesDto[];
  asset_categories: AssetCategoryDto[];
  assets: AssetDto[];
  search_engines: SearchEngineDto[];
}

/**
 * 服务器返回的完整同步数据结构，与 Rust 中的 `ServerSyncData` 对应。
 */
export interface ServerSyncData {
  current_synced_at: string;
  sync_data: SyncDataDto;
  icons_to_upload: string[]; // 需要客户端上传的图标文件名列表
  icons_to_download: string[]; // 需要客户端下载的图标文件名列表
  website_groups_count: number;
  websites_count: number;
  categories_count: number;
  assets_count: number;
  search_engines_count: number;
}

/**
 * @interface ClientSyncPayload
 * @description 客户端发起同步时，`sync/start` 接口的请求体。
 */
export interface ClientSyncPayload {
  user_uuid: string;
  last_synced_at: string;
}

/**
 * @interface StartSyncResponse
 * @description `sync/start` 接口成功响应后，服务器返回的数据结构。
 */
export interface StartSyncResponse {
  session_id: string;
  // 未来可扩展字段，例如服务器建议的分块大小等
}

/**
 * @enum DataType
 * @description 定义了同步数据块的类型，确保客户端和服务端使用一致的标识符。
 */
export enum DataType {
  WebsiteGroups = "WebsiteGroups",
  Websites = "Websites",
  AssetCategories = "AssetCategories",
  Assets = "Assets",
  SearchEngines = "SearchEngines",
  LocalIcons = "LocalIcons",
}

/**
 * @interface ClientSyncDataChunk
 * @description 客户端向 `sync/chunk` 接口发送的单个数据块的结构。
 */
export interface ClientSyncDataChunk {
  session_id: string;
  data_type: DataType;
  chunk_data:
    | WebsiteGroupDto[]
    | WebsitesDto[]
    | AssetCategoryDto[]
    | AssetDto[]
    | SearchEngineDto[]
    | string[];
}

export interface SyncStatusUpdaters {
  setIsSyncing: (isSyncing: boolean) => void;
  setSyncMessage: (message: string) => void;
  setSyncCompleted: (completed: boolean) => void;
  incrementDataVersion: () => void;
  switchActiveUser: (user: User) => void;
  refreshAvailableUsers: () => Promise<User[]>;
  // 将 t 函数作为参数传入
  t: (key: string, options?: any) => string;
}

export interface CurrentUserPayload {
  username: string;
}

/**
 * @brief 客户端信息。
 */
export interface ClientInfoDto {
  app_version: string;
  username: string;
  token: string;
  server_address: string;
}

File: src-tauri/src/types/sync.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

/// 服务器接口响应类型，与服务器中定义的数据结构一致
#[derive(Serialize, Deserialize, Debug)]
pub struct ApiResponse<T> {
    pub success: bool,
    pub code: i32,
    pub message: String,
    pub data: Option<T>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct WebsiteGroupDto {
    pub uuid: String,
    pub name: String,
    pub description: Option<String>,
    pub sort_order: Option<i64>, // Rust 中对应 SQLite 的 INTEGER
    pub is_deleted: i64,         // 对应 SQLite 的 INTEGER (0 或 1)
    pub rev: i64,
    pub updated_at: String,      // ISO 8601 格式的字符串
}

#[derive(Serialize, Deserialize, Debug)]
pub struct WebsitesDto {
    pub uuid: String,
    pub group_uuid: String, // 关系字段必须保留
    pub title: String,
    pub url: String,
    pub url_lan: Option<String>,
    pub default_icon: Option<String>,
    pub local_icon_path: Option<String>,
    pub background_color: Option<String>,
    pub description: Option<String>,
    pub sort_order: Option<i64>,
    pub is_deleted: i64,
    pub rev: i64,
    pub updated_at: String,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct AssetCategoryDto {
    pub uuid: String,
    pub name: String,
    pub is_default: i64, // 服务器可能需要根据这个标志做特殊处理
    pub is_deleted: i64,
    pub rev: i64,
    pub updated_at: String,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct AssetDto {
    pub uuid: String,
    pub category_uuid: String, // 关系字段必须保留
    pub name: String,
    pub purchase_date: String,
    pub price: f64, // 对应 SQLite 的 REAL
    pub expiration_date: Option<String>,
    pub description: Option<String>,
    pub is_deleted: i64,
    pub rev: i64,
    pub updated_at: String,
    pub brand: Option<String>,
    pub model: Option<String>,
    pub serial_number: Option<String>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct SearchEngineDto {
    pub uuid: String,
    pub name: String,
    pub url_template: String,
    pub default_icon: Option<String>,
    pub local_icon_path: Option<String>,
    pub is_default: i64,
    pub sort_order: Option<i64>,
    pub is_deleted: i64,
    pub rev: i64,
    pub updated_at: String,
}

// 定义同步数据的结构
#[derive(Serialize, Deserialize, Debug)]
pub struct SyncDataDto {
    pub website_groups: Vec<WebsiteGroupDto>,
    pub websites: Vec<WebsitesDto>,
    pub asset_categories: Vec<AssetCategoryDto>,
    pub assets: Vec<AssetDto>,
    pub search_engines: Vec<SearchEngineDto>,
}

/// 服务端成功同步后返回的数据
#[derive(Debug, Serialize, Deserialize)]
pub struct ServerSyncData {
    pub current_synced_rev: i64,
    pub current_synced_at: String,
    pub sync_data: SyncDataDto,
    pub icons_to_upload: Vec<String>, // 需要客户端上传的图标文件名列表
    pub icons_to_download: Vec<String>, // 需要客户端下载的图标文件名列表
    pub website_groups_count: usize,
    pub websites_count: usize,
    pub categories_count: usize,
    pub assets_count: usize,
    pub search_engines_count: usize,
}

/// 定义从客户端发送的数据结构
#[derive(Serialize, Deserialize, Debug)]
pub struct ClientSyncData {
    pub user_uuid: String, // 用户UUID，只发送一次
    pub last_synced_rev: i64,
    pub sync_data: SyncDataDto,
    pub local_icons: Vec<String>, // 客户端本地拥有的所有图标文件名
}

/// 定义客户端信息
#[derive(Serialize, Deserialize, Debug)]
pub struct ClientInfoDto {
    pub app_version: String,
    pub username: String,
    pub token: String,
    pub server_address: String,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct VersionInfo {
    pub version: String,
    pub commit_hash: String,
    pub build_time: String,
    pub build_env: String,
}

/// `sync/start` 接口的请求体
#[derive(Serialize, Deserialize, Debug)]
pub struct ClientSyncPayload {
    pub user_uuid: String,
    pub last_synced_rev: i64,
}

/// `sync/start` 接口的成功响应数据
#[derive(Serialize, Deserialize, Debug)]
pub struct StartSyncResponse {
    pub session_id: String,
    pub suggested_chunk_size: Option<usize>,
}

/// 数据块的类型枚举
#[derive(Serialize, Deserialize, Debug)]
pub enum DataType {
    WebsiteGroups,
    Websites,
    AssetCategories,
    Assets,
    SearchEngines,
    LocalIcons,
}

/// `sync/chunk` 接口的请求体
#[derive(Serialize, Deserialize, Debug)]
pub struct ClientSyncDataChunk {
    pub session_id: String,
    pub data_type: DataType,
    // 使用 serde_json::Value 来灵活接收不同类型的数组
    pub chunk_data: Value,
}
